### PR TITLE
Tips views: create views for tips app

### PR DESCRIPTION
**Pull request by Yoavsc129**
TO-DO: tests for views
‏‏טיפסביתבסיסי
‏‏טיפסליצורבסיסי
‏‏טיפסעריכה

### CODE REVIEW

File: tips/templates/tips/detail.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Posted by {{tip.author}} at {{tip.date}} in {{ tip.category }}
+<h1>{{ tip.text }}</h1>
+<li><a href="/tips/">return</a></li>
```
Comment (shlior7) [inline tips/templates/tips/detail.html:3]: I think you should put all the tips in one place and not as link 
the person looking at that links list have no idea what every tip is

lets set up all the tips in a sort of gallery of tips like a notes gallery

File: tips/migrations/0001_initial.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 4.0.3 on 2022-04-14 08:58
```
Comment (ShellyGolden) [inline tips/migrations/0001_initial.py:1]: You already have this file in https://github.com/redhat-beyond/Xpense/pull/49 PR, 
why you created it again?

File: tips/models.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from django.db import models
```
Comment (ShellyGolden) [inline tips/models.py:1]: Also this file created

File: tips/templates/tips/board.html [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  <head>
+    <title>Tips</title>
+	<style type="text/css">
+	  .author, .eom { color: grey }
+	</style>
+  </head>
+  <body>
+    <h1>Tips</h1>
+<li><a href="add_tip/">Add new tip</a></li>
+{% for tip in tips %}
```
Comment (ShellyGolden) [inline tips/templates/tips/board.html:12]: I think that it will be better if it will be possible to see the tips and not as a link. 
I saw this page line boars with notes in my mind

**Review by ShellyGolden — COMMENTED** [review]
1. Why you didn't work on Or's PR? 
2. Maybe the option to add tip can be a button

**Review by ShellyGolden — COMMENTED** [review]
maybe the option to add tip can be a button

File: tips/templates/tips/detail.html (hunk 1 above)
Comment (Yoavsc129) [reply tips/templates/tips/detail.html:3]: The idea is to display the tips content in a different page and in the main page show either a short snippet of the text or a separate title

Comment (shlior7) [reply tips/templates/tips/detail.html:3]: image

Comment (ronyyosef) [reply tips/templates/tips/detail.html:3]: I agree with Lior
As far as I see the tips model doesn't have a "short snippet" of the text as you said.
If we want one, we need to add it to the model and also let the user add that data in the form
I think that the view here should be some kind of an Infinite Scroll
What do you think about this idea?

Comment (Yoavsc129) [reply tips/templates/tips/detail.html:3]: It's a good idea, I'll look for a good way to add it

Comment (Yoavsc129) [reply tips/templates/tips/detail.html:3]: It's a good idea, I'll look for a good way to add it

File: tips/urls.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+
+from tips import views
+
+app_name = 'tips'
+urlpatterns = [
+    path('', views.board, name='board'),
+    path('tips/<int:tip_id>/', views.detail, name='detail'),
```
Comment (ronyyosef) [inline tips/urls.py:8]: tips/tips/id
that's a bit weird, I think you should change to a bit clear different URLs

**Review by ronyyosef — CHANGES_REQUESTED** [review]

File: tips/urls.py (hunk 5 above)
Comment (Yoavsc129) [reply tips/urls.py:8]: Will be fixed

**Comment by ronyyosef** [comment]
The html file location of the templates  files here is diffrent from other PRs please put the templates files in the templates folder

**Comment by Yoavsc129** [comment]
Added an option to edit tips

File: templates/tips/board.html [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  </head>
+  <body>
+    <h1>Tips</h1>
+	<a class="btn btn-primary" href="add_tip/" role="button">Add new tip</a>
+    <div class="grid" >
+      {% for tip in tips %}
+      <div class="item">
+        <div class="category">{{tip.category}}</div>
+        <h6 class="author">author: {{ tip.author }}</h6>
+        <label class="date">date: {{ tip.date }}</label>
```
Comment (LiorKGOW) [inline templates/tips/board.html:52]: Is it a last edited date? or a publish date?
I think it should mention whether it's the former or the latter

File: templates/tips/detail.html [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{% extends "base.html" %} {%load static%} {%block content%}
+Posted by {{tip.author}} at {{tip.date}} in {{ tip.category }}
```
Comment (LiorKGOW) [inline templates/tips/detail.html:2]: From this line I understand the date is for the date the Tip was first published
If we want to add the option to edit of a Tip, maybe it would be nice to add the last_edited date
What do you think? @Yoavsc129

**Review by LiorKGOW — CHANGES_REQUESTED** [review]

File: templates/tips/board.html (hunk 6 above)
Comment (Yoavsc129) [reply templates/tips/board.html:52]: It's always the publish date. Ill make it clearer

File: templates/tips/detail.html (hunk 7 above)
Comment (Yoavsc129) [reply templates/tips/detail.html:2]: This page (tips/detail) is not used in the latest update and I might delete it

Comment (Yoavsc129) [reply templates/tips/detail.html]: deleted

File: templates/tips/board.html (hunk 6 above)
Comment (Yoavsc129) [reply templates/tips/board.html]: fixed

File: templates/tips/board.html [hunk 8]
```diff
@@ -0,0 +1,67 @@
+{% extends "base.html" %} {%load static%} {%block content%}
+<!DOCTYPE html>
+<html lang="en">
```
Comment (ronyyosef) [inline templates/tips/board.html:3]: the HTML tag is not needed here, look at the base.html that you extend from

File: templates/tips/board.html [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  <head>
+    <title>Tips</title>
+    <style type="text/css">
```
Comment (ronyyosef) [inline templates/tips/board.html:6]: please put style in the separate file, look at what we did on the global page

File: templates/tips/board.html [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+          top:0;
+          right:0;
+      }
+      .edit{
+        position: absolute;
+        bottom:0;
+        left:0;
+      }
+    </style>
+    <script src="https://unpkg.com/packery@2/dist/packery.pkgd.min.js"></script>
```
Comment (ronyyosef) [inline templates/tips/board.html:42]: what is that script doing?

File: templates/tips/board.html [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        <div class="category">{{tip.category}}</div>
+        <h6 class="author">author: {{ tip.author }}</h6>
+		<h6 class="text">{{ tip.text }}</h6>
+        <label class="date">publish date: {{ tip.date }}</label>
+        <a class ="edit" href="edit_tip/{{ tip.id }}">Edit</a>
+      </div>
+      {% endfor %}
+    </div>
+  </body>
+  <script>
```
Comment (ronyyosef) [inline templates/tips/board.html:59]: what is that script doing?

File: templates/tips/edit_tip.html [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{% extends "base.html" %} {%load static%} {%block content%}
+<div class="editfrm">
+	<p class="edittip"><h1>Edit tip</h1></p>
+	<form method="POST">
+		{% csrf_token %}
+		{{ form.as_p }}
+		<input type="submit" value="Post">
```
Comment (ronyyosef) [inline templates/tips/edit_tip.html:7]: Do you show the current version of the tip

Comment (Yoavsc129) [reply templates/tips/edit_tip.html:7]: It does!

File: templates/tips/board.html (hunk 10 above)
Comment (Yoavsc129) [reply templates/tips/board.html:42]: For the grid layout

File: templates/tips/board.html (hunk 11 above)
Comment (Yoavsc129) [reply templates/tips/board.html:59]: It's unnecessary and will be removed

File: templates/tips/board.html (hunk 10 above)
Comment (Yoavsc129) [reply templates/tips/board.html:42]: actually, i'm not sure if this is necessary either and I will remove it

File: static/css/tips_board.css [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+          position: absolute;
+          top:0;
+          right:0;
+      }
+      .edit{
+        position: absolute;
+        bottom:0;
+        left:0;
+      }
+    </style>
```
Comment (ShellyGolden) [inline static/css/tips_board.css:36]: you need to add a new line at the end of the file

**Review by ShellyGolden — COMMENTED** [review]
Maybe it will be nice to add file CONSTANTS like we had in the house model (you can see what Lior did and I added :) )

**Comment by LiorKGOW** [comment]
#77 was merged
pls rebase

File: tips/templates/tips/detail.html (hunk 1 above)
Comment (Yoavsc129) [reply tips/templates/tips/detail.html]: comments about visual design should move here https://github.com/redhat-beyond/Xpense/pull/78

File: static/css/tips_board.css (hunk 13 above)
Comment (Yoavsc129) [reply static/css/tips_board.css]: fixed

File: templates/base.html [hunk 14]
```diff
@@ -52,6 +52,9 @@
             </li>
             <li class="nav-item">
               <a class="nav-link" href="/Mine">Mine</a>
+            </li>
+			<li class="nav-item">
```
Comment (LiorKGOW) [inline templates/base.html:56]: Align it with the rest pls
It looks better this way :)

**Review by LiorKGOW — CHANGES_REQUESTED** [review]

File: templates/base.html (hunk 14 above)
Comment (Yoavsc129) [reply templates/base.html:56]: what's not aliigned?

Comment (Yoavsc129) [reply templates/base.html]: figured it out, tabs are a menace

**Comment by ronyyosef** [comment]
Looks very good @Yoavsc129 
Is the first in this PR relvent?  commit "a96f6e81eb24abcea1ecefa0fa918f4d9151fabe"

**Review by ronyyosef — APPROVED** [review]
Well done!
I think its ready for merge

File: tips/tests/tips_views_tests.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        try:
+            get_template('tips/add_tip.html')
+        except TemplateDoesNotExist:
+            assert False, f"Template {'tips/add_tip.html'} does not exist"
+        try:
+            get_template('tips/edit_tip.html')
+        except TemplateDoesNotExist:
+            assert False, f"Template {'tips/edit_tip.html'} does not exist"
+
+    def test_edit_tip(self, generate_form, generate_tip):
```
Comment (ShellyGolden) [inline tips/tests/tips_views_tests.py:77]: what about testing add tip?

Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py]: sure, why not

File: tips/tests/tips_views_tests.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        generate_tip.save()
+        tip = get_object_or_404(Tip, pk=generate_tip.id)
+        assert tip.author == AUTHOR
+        form = generate_form
+        tip = form.save(commit=False)
+        tip.save()
+        assert tip.author == OTHER_AUTHOR
+
```
Comment (adiabramovitch) [inline tips/tests/tips_views_tests.py:67]: What happens if you edit a Tip and you delete the description? Does it work? You should add a test to validate that.

File: tips/tests/tips_views_tests.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+@pytest.mark.django_db
+class TestTipForms:
+    def test_form(self, db, generate_form, generate_tip):
+        form = generate_form
+        assert form.is_valid()
+
+    def test_add_tip(self, generate_form):
+        form = generate_form
+        tip = form.save()
```
Comment (adiabramovitch) [inline tips/tests/tips_views_tests.py:56]: What happens if you add a Tip and you don't fill the description? Does it work? You should add a test to validate that.

File: templates/tips/edit_tip.html [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+	</form>
+</div>
+<a class="btn btn-primary" href="/tips/" role="button">cancel</a>
```
Comment (adiabramovitch) [inline templates/tips/edit_tip.html:10]: 'cancel' should be uppercased

File: tips/tests/tips_views_tests.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+from tips.views import board
+from django.template.loader import get_template
+
+AUTHOR = 'RICK'
+OTHER_AUTHOR = 'RICKEN'
+CATEGORY = 'Clothing'
+DATE = timezone.now()
+TEXT = 'I BOUGHT A TESLA IN 50$ DISCOUNT IN LEVINSKI MARKET'
+
+FORM_DATA = {'author': OTHER_AUTHOR, 'text': TEXT, 'category': CATEGORY}
```
Comment (adiabramovitch) [inline tips/tests/tips_views_tests.py:17]: When writing a test we want to check bad data as well. we are checking for edge cases.

File: tips/tests/tips_views_tests.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+        form = generate_form
+        tip = form.save(commit=False)
+        tip.save()
+        assert tip.author == OTHER_AUTHOR
+
+
+@pytest.mark.django_db
+class TestTipViews:
+    def test_views_functions_200(self, generate_get_request, generate_tip):
+        try:
```
Comment (adiabramovitch) [inline tips/tests/tips_views_tests.py:72]: What does that name mean? what are you trying to check here? it should be very clear from the name.
for example: if you are checking the board function, it should be included in the test name.

File: tips/tests/tips_views_tests.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+            render = board(generate_get_request)
+            assert render.status_code == 200, "status code is not 200"
+        except Exception:
+            assert False, "Error in board function"
+
+    def test_views_templates(self):
+        try:
```
Comment (adiabramovitch) [inline tips/tests/tips_views_tests.py:79]: which templates are you testing for?  "tips" is not included in the name

File: tips/views.py [hunk 22]
```diff
@@ -1,2 +1,34 @@
...
+    tip = get_object_or_404(Tip, pk=tip_id)
+    form = TipForm(request.POST or None, instance=tip)
+    context = {'form': form}
+    if request.method == "POST":
+        if form.is_valid():
+            tip = form.save(commit=False)
+            tip.save()
+            context = {'form': form}
+            return redirect('/tips')
+    else:
```
Comment (adiabramovitch) [inline tips/views.py:32]: What about deleting a Tip?

File: templates/tips/add_tip.html [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+<div class="newfrm">
+	<p class="newtip"><h1>New tip</h1></p>
+	<form method="POST">
+		{% csrf_token %}
+		{{ form.as_p }}
+		<input type="submit" value="Post">
+	</form>
+</div>
+<a class="btn btn-primary" href="/tips/" role="button">cancel</a>
+{% endblock %}
```
Comment (adiabramovitch) [inline templates/tips/add_tip.html:11]: 'cancel' should be uppercased

**Comment by adiabramovitch** [comment]
@Yoavsc129 Good Job overall! 
I left some comments if you can address them. 

- What do we want the UI to look like? Is that a draft just for the demo to show the functionality Or this is the UI you have decided on?  
- What about filtering the tips based on their category? 
- I would make a distinction between the author name and category to the Tip itself. (maybe shift the author name to the left and change the size, font color whatever you think is right) Same for the date.

**Comment by Yoavsc129** [comment]
> @Yoavsc129 Good Job overall! I left some comments if you can address them.
> 
> * What do we want the UI to look like? Is that a draft just for the demo to show the functionality Or this is the UI you have decided on?
> * What about filtering the tips based on their category?
> * I would make a distinction between the author name and category to the Tip itself. (maybe shift the author name to the left and change the size, font color whatever you think is right) Same for the date.

This PR is mostly for the basic functionality. The UI and the filter is being worked on in https://github.com/redhat-beyond/Xpense/pull/78

File: templates/tips/add_tip.html (hunk 23 above)
Comment (Yoavsc129) [reply templates/tips/add_tip.html]: done

File: tips/views.py (hunk 22 above)
Comment (Yoavsc129) [reply tips/views.py:32]: done

File: tips/tests/tips_views_tests.py (hunk 21 above)
Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py]: done

File: tips/tests/tips_views_tests.py (hunk 20 above)
Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py]: done

File: tips/tests/tips_views_tests.py (hunk 19 above)
Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py:17]: done

File: templates/tips/edit_tip.html (hunk 18 above)
Comment (Yoavsc129) [reply templates/tips/edit_tip.html]: done

File: tips/tests/tips_views_tests.py (hunk 17 above)
Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py:64]: done

File: tips/tests/tips_views_tests.py (hunk 16 above)
Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py:75]: It's like in add_tip because the form is used

**Comment by LiorKGOW** [comment]
> Added an option to edit tip

Could you add a screenshot to show how the edit page looks?

File: tips/views.py [hunk 24]
```diff
@@ -1,2 +1,40 @@
-# from django.shortcuts import render
-# Create your views here.
+from django.shortcuts import render, redirect, get_object_or_404
+from .models import Tip
+from .forms import TipForm
+
+
+def board(request):
+    tips = Tip.objects.order_by('-date')
+    return render(request, 'tips/board.html', {'tips': tips})
```
Comment (LiorKGOW) [inline tips/views.py:8]: You should add a constant URL, like what we did on house, check out our views.py and constants.py, every PR about House should have it.
I think it would make the code a bit nicer :)

File: tips/views.py [hunk 25]
```diff
@@ -1,2 +1,40 @@
...
+
+def add_tip(request):
+    if request.method == "POST":
+        form = TipForm(request.POST)
+        if form.is_valid():
+            form.save()
+            return redirect('/tips')
+    else:
+        form = TipForm()
+    return render(request, 'tips/add_tip.html', {'form': form})
```
Comment (LiorKGOW) [inline tips/views.py:19]: same here ⬆️

File: tips/views.py [hunk 26]
```diff
@@ -1,2 +1,40 @@
...
+    context = {'form': form}
+    if request.method == "POST":
+        if form.is_valid():
+            tip = form.save(commit=False)
+            tip.save()
+            context = {'form': form}
+            return redirect('/tips')
+    else:
+        form = TipForm()
+    return render(request, 'tips/edit_tip.html', context)
```
Comment (LiorKGOW) [inline tips/views.py:34]: same here ⬆️

**Review by LiorKGOW — CHANGES_REQUESTED** [review]
Nice work @Yoavsc129 
I have dropped you some comments 👍

**Comment by LiorKGOW** [comment]
> > > Added an option to edit tip
> > 
> > 
> > Could you add a screenshot to show how the edit page looks?
> 
> you can see it in the top comment @LiorKGOW

Oh I guess I misread it thanks @ShellyGolden

File: tips/tests/tips_views_tests.py [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+            render = board(generate_get_request)
+            assert render.status_code == 200, "status code is not 200"
+        except Exception:
+            assert False, "Error in board function"
+
+    def test_tips_views_templates(self):
+        try:
+            get_template('tips/board.html')
+        except TemplateDoesNotExist:
+            assert False, f"Template {'tips/board.html'} does not exist"
```
Comment (ShellyGolden) [inline tips/tests/tips_views_tests.py:94]: what about adding constants?

Comment (Yoavsc129) [reply tips/tests/tips_views_tests.py]: sure

File: tips/views.py (hunk 24 above)
Comment (Yoavsc129) [reply tips/views.py]: sure

**Review by ShellyGolden — APPROVED** [review]
looks good!
well done :)

**Review by ronyyosef — APPROVED** [review]

**Comment by adiabramovitch** [comment]
What about the third point I raised? Is it solved in the UI PR as well?

**Comment by Yoavsc129** [comment]
> What about the third point I raised? Is it solved in the UI PR as well?

Yes

**Comment by adiabramovitch** [comment]
Also, can you squash the commits into one? all of them are referring to the same logic so I would leave it as one generalized commit. That way It will be easier to follow the changes in the git log history

**Comment by Yoavsc129** [comment]
done
